features/net/network-socket/WiFiInterface.h

```diff
@@ -57,15 +53,38 @@ class WiFiInterface: public NetworkInterface
 
     /** Start the interface
      *
-     *  Attempts to connect to a WiFi network. If passphrase is invalid,
-     *  NSAPI_ERROR_AUTH_ERROR is returned.
+     *  Attempts to connect to a WiFi network.
+     *
+     *  @param ssid      Name of the network to connect to
+     *  @param pass      Security passphrase to connect to the network
+     *  @param security  Type of encryption for connection (Default: NSAPI_SECURITY_NONE)
+     *  @param channel   Channel on which the connection is to be made, or 0 for any (Default: 0)
+     *  @param timeout   Timeout in milliseconds; 0 for no timeout (Default: 0)
+     *  @return          0 on success, or error code on failure
+     */
+    virtual int connect(const char *ssid, const char *pass, nsapi_security_t security = NSAPI_SECURITY_NONE,
+                                  uint8_t channel = 0, unsigned timeout = 0) = 0;
+
+    /** Start the interface
+     *
+     *  Attempts to connect to a WiFi network asynchronously, the call will return straight away. If the @a cb was NULL
+     *  you'll need to query @a get_state until it's in NSAPI_IF_STATE_CONNECTED state, otherwise the @a cb will be
+     *  called with connection results, connected AP details and user data.
+     *
+     *  Note: @a ssid and @a pass must be kept until the connection is made, that is the callback has been called or the
+     *        state changed to @a NSAPI_IF_STATE_CONNECTED as they are passed by value.
      *
      *  @param ssid      Name of the network to connect to
      *  @param pass      Security passphrase to connect to the network
-     *  @param security  Type of encryption for connection
-     *  @return          0 on success, negative error code on failure
+     *  @param channel   Channel on which the connection is to be made, or 0 for any (Default: 0)
+     *  @param security  Type of encryption for connection (Default: NSAPI_SECURITY_NONE)
+     *  @param cb        Function to be called when the connect finishes (Default: NULL)
+     *  @param data      Arbitrary user data to pass to @a cb function (Default: NULL)
+     *  @param timeout   Timeout in milliseconds; 0 for no timeout (Default: 0)
      */
-    virtual int connect(const char *ssid, const char *pass, nsapi_security_t security = NSAPI_SECURITY_NONE) = 0;
+    virtual void connect_async(const char *ssid, const char *pass,nsapi_security_t security = NSAPI_SECURITY_NONE,
+                               uint8_t channel = 0, mbed::Callback<void(nsapi_error_t, wifi_ap_t*, void*)> cb = NULL, void *data = NULL,
```

Comment:
With the callback class, the data argument is unecessary. The callback class natively supports user-context:
``` cpp
wifi.connect_async("ssid", "pass", NSAPI_SECURITY_NONE, 0, callback(&thing, &Thing::update));
```